### PR TITLE
Stronger tests + tweaks to mp3 codec

### PR DESCRIPTION
**Pull request by keepingitneil**

### CODE REVIEW

File: livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/tts.py
Comment (theomonnom) [inline livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/tts.py:73]: This isn't streamed to the users. This is waiting for the "download" to finish before yielding anything to the user

Comment (theomonnom) [reply livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/tts.py:73]: You could use asyncio.select or __anext__ after a http chunk

**Review by theomonnom — CHANGES_REQUESTED** [review]

File: livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/tts.py
Comment (keepingitneil) [reply livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/tts.py:73]: Hrmm so it isn't waiting for the download to complete. If 4096 bytes come in the decoder starts decoding them. It's possible (and probably) for a few http chunks to come in before the first decoded result but I think that's ok.

I do think decoding should happen on another thread (but maybe just one other thread instead of constantly doing `to_thread`?). I think the code-complexity tradeoff is worth it. Also regardless of internal implementation, I think the decoder api should remain push_chunk + async iterator. 

I'm assuming there will be slower decoders (video?) in the future and having a consistent api probably makes sense, especially when the only downside is a few lines of code. I'm also admittedly a bit scared of debugging issues of users running on cheap hardware decoding longer-than-anticipated audio where the symptoms manifest in non-obvious ways.

I know you've been pretty in favor of sync decoding - are there other reasons that I'm not seeing for your preference?

Comment (keepingitneil) [reply livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/tts.py:72]: > EDIT: You're creating a new task, I think this should be avoided wdyt?

Hrmm, so I don't have a strong opinion on task creation and whether or not we should try and avoid it. This is moreso a discussion about what should the mp3.StreamDecoder API look like. And _possibly_ a discussion about what should all decoders look like.

I went with push_chunk + a stream decoder mainly because it felt like it communicated that this decoder is stateful. The other option is `async decode_chunk(<some http chunk>) -> rtc.AudioFrame` I'm not totally against but it "feels" stateless even though it's not.

File: tests/change-sophie.wav
Comment (keepingitneil) [inline tests/change-sophie.wav:1]: moved to lfs

File: livekit-agents/livekit/agents/tts/stream_adapter.py
Comment (theomonnom) [inline livekit-agents/livekit/agents/tts/stream_adapter.py:46]:
```suggestion

```

File: livekit-agents/livekit/agents/tts/stream_adapter.py
Comment (theomonnom) [inline livekit-agents/livekit/agents/tts/stream_adapter.py:61]:
```suggestion
           self._event_queue.put_nowait(None)
```

File: livekit-agents/livekit/agents/codecs/mp3.py
Comment (theomonnom) [inline livekit-agents/livekit/agents/codecs/mp3.py:38]: Let's not type here, bytes and rtc.AudioFrame are wrong since you're also pushing None

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (theomonnom) [inline livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:306]: What is the motivation for this behaviour, if you're closing VAD and not sending enough data before, if the user is still speaking then the END_OF_SPEECH event not being sent is expected IMO

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (theomonnom) [inline livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:205]: Should this be hardcoded?

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (theomonnom) [inline livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:346]: Can u add a link that mentions the docs?
The event handling is becoming quite complex here and I'm not sure if UterranceEnd sufficient by himself.
e.g we're warning when there is no _final_events, maybe this will happen and this is expected

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (keepingitneil) [reply livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:205]: hrmm probably not. I think it can be Min(1000, endpointing).

> It is recommended that you set the value of utterance_end_ms to be 1000 ms or higher. This is because UtteranceEnd relies on Deepgram's interim_results feature. Deepgram's Interim Results are typically sent every one second, so using a value of less than one second for utterance_end_ms will not offer any benefits.

File: livekit-plugins/livekit-plugins-silero/livekit/plugins/silero/vad.py
Comment (keepingitneil) [reply livekit-plugins/livekit-plugins-silero/livekit/plugins/silero/vad.py:157]: The issue addressed here is that we use the VAD END_OF_SPEECH event in the stt stream adapter. In the close case END_OF_SPEECH was never getting called so the stream adapter was never running its inference.

I see your point but I can see it the other way too. What would your preferrered solution be? Collect audio frames in the stream_adapter itself?

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (keepingitneil) [inline livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:315]: @theomonnom I removed the return in this case so that a START_OF_SPEECH event is matched with an END_OF_SPEECH event even if it's likely a bug if we hit this condition

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (keepingitneil) [reply livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:357]: Made some adjustments, lmk if it makes sense. This logic is definitely confusing - It could be worth writing a custom test just for deepgram because they have two "VAD" systems that work differently. One is the endpointing and it's pure DSP. Background noise will trigger it. the other is utterance end which is more NLP based.

File: tests/test_stt.py
Comment (keepingitneil) [reply tests/test_stt.py:82]: Removed the short test constraint because it was passing. But even though this is threshold looks agressive, it's not really:

```
=========================== short test summary info ============================
FAILED tests/test_stt.py::test_recognize[google-long] - assert 0.164634[146](https://github.com/livekit/agents/actions/runs/8160618138/job/22307572857#step:6:147)34146342 > 0.5
 +  where 0.16463414634146342 = <bound method SequenceMatcher.ratio of <difflib.SequenceMatcher object at 0x7f8cdf75d4c0>>()
 +    where <bound method SequenceMatcher.ratio of <difflib.SequenceMatcher object at 0x7f8cdf75d4c0>> = <difflib.SequenceMatcher object at 0x7f8cdf75d4c0>.ratio
 +      where <difflib.SequenceMatcher object at 0x7f8cdf75d4c0> = SequenceMatcher(None, "My girlfriend is asleep. So I can't talk loud, but that's probably pretty good for this test. This is a long test for", "My girlfriend is asleep so I can't talk loud, but that's probably pretty good for this test. This is a long test for speech-to-text, it has some pauses in it, it might have some background noise in it, we'll see")
FAILED tests/test_stt.py::test_stream[google-long] - assert 0.14594594594594595 > 0.5
 +  where 0.14594594594594595 = <bound method SequenceMatcher.ratio of <difflib.SequenceMatcher object at 0x7f8cd4f50050>>()
 +    where <bound method SequenceMatcher.ratio of <difflib.SequenceMatcher object at 0x7f8cd4f50050>> = <difflib.SequenceMatcher object at 0x7f8cd4f50050>.ratio
 +      where <difflib.SequenceMatcher object at 0x7f8cd4f50050> = SequenceMatcher(None, " My girlfriend is asleep. So I can't talk loud, but that's probably pretty good for this test. This is a long test for Speech to text it has some pauses in it.", "My girlfriend is asleep so I can't talk loud, but that's probably pretty good for this test. This is a long test for speech-to-text, it has some pauses in it, it might have some background noise in it, we'll see")
============= 2 failed, 14 passed, 4 warnings in 111.71s (0:01:51) =============
```

Our diffing algorithm isn't very strong. This is to me looks like >0.16 match

File: livekit-plugins/livekit-plugins-silero/livekit/plugins/silero/vad.py
Comment (keepingitneil) [reply livekit-plugins/livekit-plugins-silero/livekit/plugins/silero/vad.py]: I removed this and added silence to the end of the other audio file. PR was too big and this is a different existing issue that can be addressed in a future PR

File: livekit-agents/livekit/agents/codecs/mp3.py
Comment (theomonnom) [inline livekit-agents/livekit/agents/codecs/mp3.py:76]: This isn't valid, you're stil using the frames with nchannels > 1
I think making the API sync will makes everything much simpler. Error handling is starting to be hard here no?

File: livekit-plugins/livekit-plugins-elevenlabs/livekit/plugins/elevenlabs/tts.py
Comment (theomonnom) [inline livekit-plugins/livekit-plugins-elevenlabs/livekit/plugins/elevenlabs/tts.py:104]: This remains unchanged inside the base class
function signatures are not consistent

File: tests/utils.py
Comment (theomonnom) [inline tests/utils.py:5]: The order isn't compared?

File: tests/test_stt.py
Comment (theomonnom) [inline tests/test_stt.py:32]: API feels a little bit weird, what does it means to close?

**Review by theomonnom — APPROVED** [review]
ok to merge once these are resolved.
but we still need to think before merging with main the API

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (theomonnom) [inline livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:358]: Can UtteranceEnd happens without SpeechStart?

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (theomonnom) [inline livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:307]:
```suggestion
    def _end_speech(self) -> None:
```

File: tests/utils.py
Comment (keepingitneil) [reply tests/utils.py:5]: nah, in the future we can do a stronger vector based approach if we need it but I think comparing lexicons is strong enough for now

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (keepingitneil) [reply livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:358]: Shouldn't, unless you receive transcribed words without a speech start. Utterance end is based on transcribed words timing:
https://developers.deepgram.com/docs/understanding-end-of-speech-detection#using-utteranceend

File: livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py
Comment (keepingitneil) [reply livekit-plugins/livekit-plugins-deepgram/livekit/plugins/deepgram/stt.py:347]: I think we can keep until we have more confidence that this logic is correct and aligned with how deepgram works

File: livekit-agents/livekit/agents/codecs/mp3.py
Comment (keepingitneil) [reply livekit-agents/livekit/agents/codecs/mp3.py:76]: fair, i'll just make it all sync in a fast-follow PR

**Pull request merged**
